File: blog/php-vs-go/interface/interface.go
/*
* @Author: GeekWho
* @Date:   2019-01-24 15:46:29
* @Last Modified by:   GeekWho
* @Last Modified time: 2019-01-24 22:27:17
*/
package main

import (
    "fmt"
)
// 申明接口有2个方法
type Switchs interface {
    On() string
    Off() string
}

// 申明一个结构体，类似PHP的Light类
type Light struct {
    Status string
}
//添加一个On方法
func (l Light) On() string {
    l.Status = "light is on"
    return l.Status
}
//添加一个Off方法
func (l Light) Off() string {
    l.Status = "light is off"
    return l.Status
}

func main() {
    // 初始化类
    l := Light{}
    fmt.Println(l.On())
    fmt.Println(l.Off())
}

File: blog/php-vs-go/interface/interface.php
<?php

/**
 * @Author: GeekWho
 * @Date:   2019-01-24 15:45:49
 * @Last Modified by:   GeekWho
 * @Last Modified time: 2019-01-24 15:46:14
 */
interface switchs {
    public function on();
    public function off();
}
class Light implements switchs
{
    public $status;
    public function __construct(){
    }
    public function on()
    {
        $this->status = 'light is on';
        echo $this->status . PHP_EOL;
    }
    public function off()
    {
        $this->status = 'light is off';
        echo $this->status . PHP_EOL;
    }
}
$light = new Light();
$light->on();
$light->off();
